File: bindings/java/com/statespec/backend/runtime/EntityGcRegistration.java
package com.statespec.backend.runtime;

import com.statespec.backend.Backend;
import com.statespec.backend.Json;
import java.util.ArrayList;
import java.util.List;

public final class EntityGcRegistration
{
    private EntityGcRegistration() {}

    @FunctionalInterface
    public interface EntityGcTask {
        void run(String now) throws Backend.BackendException;
    }

    @FunctionalInterface
    public interface Registrar {
        void add(EntityGcTask task) throws Backend.BackendException;
    }

    public static final class DefaultEntityGcRepository implements EntityGcRepository
    {
        @Override
        public List<Backend.VersionedRecord> listGcEligibleTx(
            Backend.Transaction tx,
            EligibilityRequest request
        ) throws Backend.BackendException
        {
            int limit = request.limit() > 0 ? request.limit() : 100;
            List<Backend.VersionedRecord> eligible = new ArrayList<>();
            for (var record : tx.query(request.descriptor().collection(), new Backend.Query.All()))
            {
                if (eligible.size() >= limit)
                {
                    break;
                }
                var status = record.document().find(request.descriptor().statusField());
                if (status.isEmpty() || !(status.get() instanceof Json.StringValue value))
                {
                    continue;
                }
                if (isTerminalGcState(request.descriptor(), value.value()))
                {
                    eligible.add(record);
                }
            }
            return eligible;
        }

        @Override
        public void finalizeGcTx(
            Backend.Transaction tx,
            FinalizeRequest request
        ) throws Backend.BackendException
        {
            tx.erase(request.descriptor().collection(), request.key());
        }
    }

    public static void registerEntityGcWorkers(
        Registrar registrar,
        Backend backend
    ) throws Backend.BackendException
    {
        registerEntityGcWorkers(registrar, backend, List.of());
    }

    public static void registerEntityGcWorkers(
        Registrar registrar,
        Backend backend,
        List<EntityGcTypes.Descriptor> descriptors
    ) throws Backend.BackendException
    {
        for (var descriptor : descriptors)
        {
            var worker = new EntityGcWorkers.Worker(
                descriptor, new DefaultEntityGcRepository(), new EntityGcWorkers.WorkerConfig(0)
            );
            registrar.add(now -> worker.runOnce(backend, now));
        }
    }

    private static boolean isTerminalGcState(
        EntityGcTypes.Descriptor descriptor,
        String status
    )
    {
        for (var terminal : descriptor.terminalStates())
        {
            if (terminal.state().equals(status))
            {
                return true;
            }
        }
        return false;
    }
}

File: bindings/rust/runtime/entity_gc_registration.rs
use std::sync::Arc;

use crate::backend::{Backend, BackendResult, Query, Transaction, VersionedRecord};
use crate::json::Json;
use crate::runtime_entity_gc_repository::{
    EntityGcEligibilityRequest, EntityGcFinalizeRequest, EntityGcRepository,
};
use crate::runtime_entity_gc_types::EntityGcDescriptor;
use crate::runtime_entity_gc_workers::{EntityGcWorker, EntityGcWorkerConfig};

#[derive(Debug, Clone, Default)]
pub struct DefaultEntityGcRepository;

impl<Tx: Transaction> EntityGcRepository<Tx> for DefaultEntityGcRepository {
    fn list_gc_eligible_tx(
        &self,
        tx: &mut Tx,
        request: &EntityGcEligibilityRequest,
    ) -> BackendResult<Vec<VersionedRecord>> {
        let limit = if request.limit == 0 { 100 } else { request.limit };
        let mut eligible = Vec::new();
        for record in tx.query(&request.descriptor.collection, &Query::All)? {
            if eligible.len() >= limit {
                break;
            }
            let Some(Json::String(status)) = record.document.find(&request.descriptor.status_field)
            else {
                continue;
            };
            if is_terminal_gc_state(&request.descriptor, status) {
                eligible.push(record);
            }
        }
        Ok(eligible)
    }

    fn finalize_gc_tx(
        &self,
        tx: &mut Tx,
        request: &EntityGcFinalizeRequest,
    ) -> BackendResult<()> {
        tx.erase(&request.descriptor.collection, &request.key)
    }
}

pub type EntityGcTask = Arc<dyn Fn(String) -> BackendResult<()> + Send + Sync>;

pub fn register_entity_gc_workers<B, F>(
    mut register: F,
    backend: Arc<B>,
    descriptors: Vec<EntityGcDescriptor>,
) -> BackendResult<()>
where
    B: Backend + Send + Sync + 'static,
    F: FnMut(EntityGcTask) -> BackendResult<()>,
{
    for descriptor in descriptors {
        let backend = Arc::clone(&backend);
        let worker = EntityGcWorker::new(
            descriptor,
            DefaultEntityGcRepository,
            EntityGcWorkerConfig::default(),
        );
        register(Arc::new(move |now| {
            worker.run_once(backend.as_ref(), &now).map(|_| ())
        }))?;
    }
    Ok(())
}

fn is_terminal_gc_state(descriptor: &EntityGcDescriptor, status: &str) -> bool {
    descriptor
        .terminal_states
        .iter()
        .any(|terminal| terminal.state == status)
}
